File: Day #613/solution.cpp
// Day 613: PrefixMapSum - insert(key,value) and sum(prefix).
// Approach: trie where each node stores total of values passing through; insert propagates delta. Time O(|key|).
#include <bits/stdc++.h>
using namespace std;

class PrefixMapSum {
    struct Node {
        long long sum = 0;
        unordered_map<char, Node*> ch;
    };
    Node* root = new Node();
    unordered_map<string, long long> values;
public:
    void insert(const string& key, long long value) {
        long long delta = value - values[key];
        values[key] = value;
        Node* node = root;
        for (char c : key) {
            if (!node->ch.count(c)) node->ch[c] = new Node();
            node = node->ch[c];
            node->sum += delta;
        }
    }
    long long sum(const string& prefix) {
        Node* node = root;
        for (char c : prefix) {
            if (!node->ch.count(c)) return 0;
            node = node->ch[c];
        }
        return node->sum;
    }
};

int main() {
    PrefixMapSum m;
    m.insert("columnar", 3);
    cout << m.sum("col") << "\n"; // 3
    m.insert("column", 2);
    cout << m.sum("col") << "\n"; // 5
    return 0;
}

File: Day #613/solution.js
// Day 613: PrefixMapSum - insert(key,value) and sum(prefix).
// Approach: trie where each node stores total of values passing through; insert propagates delta. Time O(|key|).
'use strict';

class Node {
    constructor() {
        this.sum = 0;
        this.ch = new Map();
    }
}

class PrefixMapSum {
    constructor() {
        this.root = new Node();
        this.values = new Map();
    }
    insert(key, value) {
        const delta = value - (this.values.get(key) || 0);
        this.values.set(key, value);
        let node = this.root;
        for (const c of key) {
            if (!node.ch.has(c)) node.ch.set(c, new Node());
            node = node.ch.get(c);
            node.sum += delta;
        }
    }
    sum(prefix) {
        let node = this.root;
        for (const c of prefix) {
            if (!node.ch.has(c)) return 0;
            node = node.ch.get(c);
        }
        return node.sum;
    }
}

function main() {
    const m = new PrefixMapSum();
    m.insert("columnar", 3);
    console.log(m.sum("col")); // 3
    m.insert("column", 2);
    console.log(m.sum("col")); // 5
}

main();
